### DISCUSSION TITLE
Middleware

### DISCUSSION THREAD

**Discussion by clayrisser** [onejs/one#218] [General]
I need to setup a websocket server, and I need access to the server instance to create it.

https://github.com/enisdenjo/graphql-ws

For example...

```ts
import * as ws from "ws";

const wsServer = new WebSocketServer({
  onestackServerInstance,
  path: '/api/graphql',
});
```

How can I access the `onestackServerInstance`?

**Comment by zetavg** [comment]
Quick reply: In your `vite.config.*`, you can add a custom plugin with a [`configureServer` hook](https://vite.dev/guide/api-plugin.html#configureserver) to access the Vite dev server instance and extend its functionality.

```js
const myServerPlugin = () => ({
  name: 'configure-server',
  configureServer(server) {
    // do something to the server
  },
})

export default defineConfig({
  plugins: [
    one(/* ... */),
    myServerPlugin(),
  ],
})
```

But I can think of a few issues that may make this unusable for your case:

* We have an internal h3 server acting as the "API gateway" for the dev server (legacy and we want to remove it), which might prevent added features from being exposed.
* What about production?

I think we should take a deeper look soon, maybe tomorrow.

**Comment by zetavg** [comment]
We're on the way of removing the internal h3 server. #210 #199

**Comment by natew** [comment]
h3 is removed now so it's all Vite.

We do need a generic way for people to hook into the server in both dev and prod. Vite doesn't provide this so maybe we should work with them a bit, but some basic middleware support would be nice. Ideally we can make it typed (Tanstack does good stuff here).
